python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -155,177 +170,272 @@ def take_table(
+    schemas: List["pyarrow.Schema"], promote_types: bool = False
+) -> "pyarrow.Schema":
+    if get_pyarrow_version() < MIN_PYARROW_VERSION_TYPE_PROMOTION:
+        return pyarrow.unify_schemas(schemas)
 
+    # NOTE: By default type promotion (from "smaller" to "larger" types) is disabled,
+    #       allowing only promotion b/w nullable and non-nullable ones
+    arrow_promote_types_mode = "permissive" if promote_types else "default"
+
+    return pyarrow.unify_schemas(schemas, promote_options=arrow_promote_types_mode)
+
+
+def unify_schemas(
+    schemas: List["pyarrow.Schema"], *, promote_types: bool = False
+) -> "pyarrow.Schema":
+    """Version of `pyarrow.unify_schemas()` which also handles checks for
+    variable-shaped tensors in the given schemas.
+
+    This function scans all input schemas to identify columns that contain
+    variable-shaped tensors or objects. For tensor columns, it ensures the
+    use of appropriate tensor types (including variable-shaped tensor types).
+    For object columns, it uses a specific object type to accommodate any
+    objects present. Additionally, it handles columns with null-typed lists
+    by determining their actual types from the given schemas.
+
+    Currently, it disallows the concatenation of tensor columns and
+    pickled object columns for performance reasons.
+    """
+    if len(schemas) == 0:
+        raise ValueError("No schemas provided for unify_schemas")
+
+    # Check if all schemas are the same
+    schemas_to_unify = [schemas[0]]
+    for schema in schemas[1:]:
+        if not schema.equals(schemas[0]):
+            schemas_to_unify.append(schema)
+
+    if len(schemas_to_unify) == 1:
+        return schemas[0]
+
+    pyarrow_result = None
+    try:
+        pyarrow_result = _pyarrow_unify_schemas(schemas_to_unify, promote_types)
+    except Exception:
+        # PyArrow failed, we'll need to use our custom handling
+        pass
+
+    # Find diverging fields - we need this whether PyArrow succeeded or not
+    divergent_fields = _find_diverging_fields(schemas_to_unify)
+
+    # Check if we need custom handling
+    needs_custom_handling = False
+    for field_info in divergent_fields.values():
+        if (
+            field_info.has_struct
+            or field_info.has_tensor
+            or field_info.has_object
+            or field_info.has_null_list
+        ):
+            needs_custom_handling = True
+            break
+
+    # If PyArrow succeeded and no custom handling needed, return its result
+    if pyarrow_result is not None and not needs_custom_handling:
+        return pyarrow_result
+
+    # Custom handling needed - reconcile divergent fields
+    schema_field_overrides = {}
+    for field_name, field_info in divergent_fields.items():
+        reconciled_type = _reconcile_field(field_info, schemas_to_unify, promote_types)
+        if reconciled_type is not None:
+            schema_field_overrides[field_name] = reconciled_type
```

Comment:
Let's drastically simplify this -- as discussed, we only need to handle Tensor types reconciliation so that's essentially the only type of conversion we need to find and reconcile (while keeping in mind that tensors could be inside structs/lists/maps as well)